src/sequencer.c

```diff
@@ -0,0 +1,193 @@
+#include "sequencer.h"
+
+typedef struct  {
+    struct event e; // parsed event -- we clear out sequence and time here obv
+    uint32_t tag;
+    uint32_t tick; // 0 means not used 
+    uint16_t divider; // 0 means not used 
+} sequence_entry;
+
+// linked list of sequence_entries
+typedef struct sequence_entry_ll_t{
+    sequence_entry *sequence;
+    struct sequence_entry_ll_t *next;
+} sequence_entry_ll_t;
+
+sequence_entry_ll_t * sequence_entry_ll_start;
+
+// Optional sequencer hook that's called every tick
+void (*amy_external_sequencer_hook)(uint32_t) = NULL;
+
+// Our internal accounting
+uint32_t sequencer_tick_count = 0;
+uint64_t next_amy_tick_us = 0;
+uint32_t us_per_tick = 0;
+
+#ifdef ESP_PLATFORM
+#include "esp_timer.h"
+esp_timer_handle_t periodic_timer;
+#else
+#include <pthread.h>
+#endif
+
+void sequencer_recompute() {
+    us_per_tick = (uint32_t) (1000000.0 / ((amy_global.tempo/60.0) * (float)AMY_SEQUENCER_PPQ));    
+    next_amy_tick_us = amy_sysclock()*1000 + us_per_tick;
+}
+
+extern int parse_list_uint32_t(char *message, uint32_t *vals, int max_num_vals, uint32_t skipped_val);
+
+// Parses the tick, divider and tag out of a sequence message
+void parse_tick_and_tag(char * message, uint32_t *tick, uint16_t *divider, uint32_t *tag) {
+    uint32_t vals[3];
+    parse_list_uint32_t(message, vals, 3, 0);
+    *tick = (uint32_t)vals[0]; *divider = (uint16_t)vals[1]; *tag = (uint32_t)vals[2];
+}
+
+uint8_t sequencer_add_event(struct event e, uint32_t tick, uint16_t divider, uint32_t tag) {
+    // add this event to the list of sequencer events in the LL
+    // if the tag already exists - if there's tick/divider, overwrite, if there's no tick / divider, we should remove the entry
+    if(divider != 0 && tick != 0) { divider = 0; } // if tick is set ignore divider 
+
+    sequence_entry_ll_t *entry_ll = sequence_entry_ll_start;
```

Comment:
sequence_entry_ll_t **entry_ll_ptr = &sequence_entry_ll_start;  // Start pointing to the root node.
while ((*entry_ll_ptr) != NULL) {
  if ((*entry_ll_ptr)->sequence->tag == tag) {
    if (divider == 0 && tick == 0) { // delete
      sequence_entry_ll_t doomed = *entry_ll_ptr;
      *entry_ll_ptr = doomed->next;  // close up list.
      free(doomed->sequence);
      free(doomed);
      return 0;
    } else { // replace
      (*entry_ll_ptr)->sequence->divider = divider;
      (*entry_ll_ptr)->sequence->tick = tick;
      (*entry_ll_ptr)->sequence->e = e;
      return 0;
    }
  }
  entry_ll_ptr = &((*entry_ll_ptr)->next);  // Update to point to the `next` field in the preceding list node.
}
// If we got here, we didn't find the tag in the list, so add it at the end.
if(tick == 0 && divider == 0) return 0;   // Ignore non-schedulable event.
if(tick != 0 && tick <= sequencer_tick_count) return 0; // don't schedule things in the past.
(*entry_ll_ptr) = malloc(sizeof(sequence_entry_ll_t));
(*entry_ll_ptr)->sequence = malloc(sizeof(sequence_entry));
(*entry_ll_ptr)->sequence->e = e;
(*entry_ll_ptr)->sequence->tick = tick;
(*entry_ll_ptr)->sequence->divider = divider;
(*entry_ll_ptr)->sequence->tag = tag;
(*entry_ll_ptr)->next = NULL;
return 1;
}